AmplifyPlugins/Geo/Sources/AWSLocationGeoPlugin/Dependency/AWSDeviceTracker.swift

```diff
@@ -35,31 +39,222 @@ class AWSDeviceTracker: NSObject, CLLocationManagerDelegate, DeviceTrackingBehav
+                switch clError {
+                case CLError.denied:
+                    // user denied location services, stop tracking
+                    stopTracking()
+                default:
+                    Amplify.log.error(error: error)
+                }
+            } else {
+                Amplify.log.error(error: error)
+            }
+    }
+    
+    public func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
+        do {
+            try checkPermissionsAndStartTracking()
+        } catch {
+            Amplify.log.error(error: error)
+        }
     }
     
     public func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
-        // TODO: Save position locally
-        proxyDelegate?.didUpdateLocations(manager, locations)
+        // if network is unreachable and `disregardLocationUpdatesWhenOffline` is set, don't store locations
+        // in local database
+        if options.disregardLocationUpdatesWhenOffline && !networkMonitor.networkConnected() {
+            return
+        }
+        
+        let currentTime = Date()
+        // check if trackUntil time has elapsed
+        if(options.trackUntil < currentTime) {
+            // flush out all stored locations in local database
+            Task {
+                do {
+                    try await sendStoredLocationsToService()
+                    try await sendReceivedLocationsToService(locations: locations, options: Geo.UpdateLocationOptions(tracker: options.tracker!))
+                } catch {
+                    // TODO: send error on Hub
+                }
+            }
+            stopTracking()
+            return
+        }
+        
+        // fetch last saved location and update time
+        var lastUpdatedLocation : CLLocation?
+        if let loadedLocation = UserDefaults.standard.data(forKey: AWSDeviceTracker.lastUpdatedLocationKey),
+           let decodedLocation = try? NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(loadedLocation) as? CLLocation {
+            lastUpdatedLocation = decodedLocation
+        }
+        let lastLocationUpdateTime = UserDefaults.standard.object(forKey: AWSDeviceTracker.lastLocationUpdateTimeKey) as? Date
+        
+        let thresholdReached = options.batchingOption._thresholdReached(
+            Geo.LocationManager.BatchingOption.LocationUpdate(timeStamp: lastLocationUpdateTime,
+                                                              position: lastUpdatedLocation),
+            Geo.LocationManager.BatchingOption.LocationUpdate(timeStamp: currentTime,
+                                                              position: locations.last)
+        )
+
+        if thresholdReached {
+            if let didUpdateLocations = options.locationProxyDelegate.didUpdateLocations {
+                do {
+                    guard let deviceId = UserDefaults.standard.object(forKey: AWSDeviceTracker.deviceIDKey) as? String else {
```

Comment:
what happens if the deviceId in UserDefaults is different than the deviceId for the locations?

Reply:
for newly received locations, it gets the deviceId from UserDefaults.
stored locations have deviceId stored with them. Currently, there is no logic to check if they're different.